Testing Initialization

Owlullaby opened the pull request:
[Trello card](https://trello.com/c/RsrLZ4oe/302-8-section-1-course-overview-development-environment-subsections-3-5)

Owlullaby committed:
feat(Section 1-4): add testing initialization project

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/.gitignore:
```diff
@@ -0,0 +1,33 @@
+# See https://help.github.com/articles/ignoring-files for more about ignoring files.
```
what is this file for?

Owlullaby replied:
`.gitignore` is used to list out files that we want to be excluded during git push. This can be useful to exclude some local files that includes private information such as `.env` file or some unrelated files that get pushed accidentally to the repository such as `.DS_Store` in Mac.

tlkiong replied:
```
some unrelated files that get pushed accidentally to the repository such as .DS_Store in Mac
```

Can you try this statement? Push a file by accident to github. Then add that file path to here. See if its ignored or not.

Owlullaby replied:
What I mean is that using `.gitignore` could prevent accidentally pushing unwanted files to the repository, but of course you have to include the path of the unwanted files inside the `.gitignore` file beforehand.

tlkiong replied:
1. You could've jz explained in this manner which is so much clearer rather than trying to sound fancy and all. Clarity is most important.
2. Since we are on this topic, did you try `Push a file by accident to github. Then add that file path to here. See if its ignored or not.`? And then, how do you ignore that file?

Owlullaby replied:
2. To push a file by accident to Github, I created a file named `unwantedfile` inside the `testing_project` directory then pushing it to git before adding in `unwantedfile` inside the `.gitignore` file.

In order to make the repository to be ignoring `unwantedfile` correctly,
run:
```
git rm --cached unwantedfile
```
then `git commit` with the correct message before `git push`,  your `.gitignore` will now works correctly.

tlkiong replied:
very good.

Now, try one more thing. Look at your `git log`. You will have a log to show you had committed that unwanted file and you can see the record of that file. How do you remove the history of that unwanted file?

Owlullaby replied:
To remove the commit history from the git log, 
first run:
```
git log --oneline
```
to show all history in one line along with the 7-digits sha1-commit-id.

In this case, since I already removed the `unwantedfile` and all I wanted is to remove the commit history with commits message starting with `test` without any further changes to my working directory and staging area, so I use `git rebase` by running:
```
git rebase -i <sha-commit-id>
```
with `<sha-commit-id>` is the sha commit id of the previous commit and `-i` command opens up the interactive mode so I can remove the two commit with `test` in the editor.

Lastly, I run:
```
git push --force-with-lease origin testing-initialization
```
to force the push and making remote branch up to date with the local branch.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/.gitattributes:
```diff
@@ -0,0 +1,10 @@
+# See https://git-scm.com/docs/gitattributes for more about git attribute files.
```
what is this file for?

Owlullaby replied:
`.gitattribute` is used to specify files and pathnames with the attributes that should be used by git when performing certain git actions. Here are some examples:
`*			text=auto` to have all files identified as text by git to have the end of line converted to `LF` i.e. `\n`
`*.txt		text` to have all files ending with `.txt` to be identified as text
`*.vcproj		text eol=crlf` to have all files ending with `.vcproj` to force to have end of line converted to `CRLF` i.e. `\r`
`*.sh		text eol=lf` to have all files ending with `.sh` to force to have end of line converted to `LF` i.e. `\n`
`*.jpg		-text` to have all files ending with `.jpg` not to be normalized as text

tlkiong replied:
So, how is this file different than `.gitignore` ?

Owlullaby replied:
`.gitignore` is used to include file path to those files that you wanted to be ignored during your git push.
`.gitattribute` on the other hand is used to give attributes to files by specifying filenames or pathnames and the given attributes.
Other than those example mentioned above, for example inside this particular `.gitattribute` file generated by Rails.
```
# Mark the database schema as having been generated.
db/schema.rb linguist-generated

# Mark the yarn lockfile as having been generated.
yarn.lock linguist-generated

# Mark any vendored files as having been vendored.
vendor/* linguist-vendored
```
`db/schema.rb` is the path to the specific file while `linguist-generated` identified the specific file as a generated code, same goes to the `yarn.lock` file.
`vendor/*` indicates that all files inside the `vendor` folder will be identified as vendored code. Vendor folder is a directory to hold third party files or codes such as plugins and gems in Rails.
Both `linguist-generated` and `linguist-vendored` indicates that these files will be excluded by Git Linguist when generating the overview of languages contained in the repository.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/Gemfile:
```diff
@@ -0,0 +1,56 @@
+source 'https://rubygems.org'
```
what is this file for?

Owlullaby replied:
`Gemfile` denotes the source url to fetch the gems package and also the gems used in the project with their respective versions.

tlkiong replied:
1. `Gemfile denotes the source url` elaborate more on this pls
2. But not everything has version number there o. How do we set the respective version?
3. Why we use this?

Owlullaby replied:
1. We put the source url: 'https://rubygems.org' so it will know which ruby server for it to fetch for the gem packages. Other than that, we can also wrap other sources in a block as below:
```
source 'https://rubygems.org'
# Gems from default source

source 'https://example.com' do
     # Gems from other source
end
```

2. When no version is specified next to the gem, then it will be updated to the newest version during `bundle update`. Below is an example of gem packages listed in a Gemfile.
```
gem 'nokogiri'
gem 'rails', '5.0.0'
gem 'rack',  '>=1.0'
gem 'thin',  '~>1.1'
```
`gem 'rails', '5.0.0'` means that we want `rails` to be specifically version 5.0.0 only.
`gem 'rack',  '>=1.0'` means that any version higher than or equal to 1.0 would work.
`gem 'thin',  '~>1.1'` means that any version higher than or equal to 1.1 would work but not 2.0 or higher. This is the same as `gem 'thin',  '>=1.1', '<2.0'`.

3. By having versions specified or using pessimistic locking `~>`, it ensures that when running `bundle update` no accidental break of code will happen.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/Gemfile.lock:
```diff
@@ -0,0 +1,230 @@
+GEM
```
what is this file for?

Owlullaby replied:
`Gemfile.lock` file is created automatically when installing project dependencies (gems) either through `bundle install` or `gem install <gems>`. It's a snapshot of all the gems and its version used that worked when running the application.

tlkiong replied:
1. "Gemfile.lock file is created automatically when installing project dependencies (gems) either through bundle install or gem install <gems>" Wrong. read again.
2. So, whats the difference between this and `Gemfile`?

Owlullaby replied:
1. `Gemfile.lock` is created automatically when installing project dependencies (gems) through `bundle install`. `gem install <gems>` installs the gems on the local machine that we are working on.
2. `Gemfile` is where we list out the gems we needed, then run `bundle install` to create a `Gemfile.lock` file so that when running the same project on another machine, it will look at the `Gemfile.lock` to know the versions for each gems the application works on rather than just getting the latest version of all the gems in the `Gemfile`.

tlkiong replied:
1. Good. Be more careful. Small details makes a complete difference in the meaning
2. Why do we need to know the version of each gem?

Owlullaby replied:
2. Running `bundle install` when `Gemfile.lock` is present would ensures that it only installs using the dependencies listed in there. Having a "lock" to specific version of each gem ensures that during deploying code to production there would not be any changes to the version of each gem that would possibly causes breaks.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/README.md:
```diff
@@ -0,0 +1,24 @@
+# README
```
what is this file for?

Owlullaby replied:
`README.md` is a markdown text file used by owner or contributors to specify more info regarding the repository. This could include info on how to work with the files, important remarks on certain changes, pictures of the working demo, license of the project and many more. This is the first thing you see displayed when you clicked into a repository.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/Rakefile:
```diff
@@ -0,0 +1,6 @@
+# Add your own tasks in files placed in lib/tasks ending in .rake,
```
what is this file for?

Owlullaby replied:
`Rake` is a task runner for Ruby. Rails provides the predefined Rake tasks for creating databases, running migrations, and performing tests. When running `rake`, it will search for `Rakefile` in the parents folder. Tasks can be written inside `Rakefile` or by adding additional rake files in the `lib/tasks` directory with file extension `.rake`.

tlkiong replied:
What is a `task runner`?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/assets/config/manifest.js:
```diff
@@ -0,0 +1,2 @@
+//= link_tree ../images
```
what is this file for?

Owlullaby replied:
`manifest.js` was created by default when running `rails new` which specify which files to be used as a top-level target using sprockets methods `link`, `link_directory`, and `link_tree`. 
In the file above, `link_tree` would include all files found inside the `/images` directory as well as its sub-directories while `link_directory` would only include all files in the parent folder, excluding files inside the sub-directories.

tlkiong replied:
Don't understand what you are saying. Layman language please

Owlullaby replied:
Starting from Rails 4 and above, `app/assets/config/manifest.js` is created on default when running `rails new`. The `manifest.js` file is needed for Sprockets used in Rails for compiling and serving web assets such as JS files, CSS files and images etc. (In Rails 6, Webpack is used for compiling JS files on default)

Inside `manifest.js`, using Sprockets directive, we could configure assets that we wanted to include during compilation for our app.

For example:
```
//= link ../javascript/application.js
//= link_directory ../stylesheets .css
//= link_tree ../images
```
`link`: include only specific file
In the example above, this would include only single file `application.js`

`link_directory`: include all files excluding the subfolders in this path
In the example above, this would include all files ending with `.css` inside the `stylesheets` directory

`link_tree`: include all files and files inside subfolders in this path
In the example above, this would include all files and files inside subfolders inside the `image` directory

tlkiong replied:
Ok. You say its needed for sprockets. Why & how is it used by sprockets?

Owlullaby replied:
Why?
Sprockets is a Ruby library for compiling and serving web assets. Sprockets allows to organize an application’s JavaScript and CSS files into smaller more manageable chunks that can be distributed over a number of directories and files. It provides structure and practices on how to include assets in our projects.

How?
Using directives inside `manifest.js` file, Sprockets can determine which files are included during compilation. When it comes to deploying your application, Sprockets then uses these directives to turn your multiple files into a single file for better performance.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/assets/stylesheets/application.css:
```diff
@@ -0,0 +1,15 @@
+/*
```
what is this file for?

Owlullaby replied:
`application.css` is where you put all your CSS styling.

tlkiong replied:
So, if you have 1 million lines of CSS, all goes into this one file?

Owlullaby replied:
You could still include additional CSS files with `.css` file extension inside the same folder if you wish to separate the CSS styling needed for your projects.

tlkiong replied:
ok. 
1. Then, how to do that?
2. What you just said is very different from what you initially state. Which is the correct one?

Owlullaby replied:
1. By using `rails g controller <controller> <view>` command, we can do:
```
rails generate controller Greetings hello
```
This would generate the following output:
```
 create  app/controllers/greetings_controller.rb
      route  get "greetings/hello"
     invoke  erb
     create    app/views/greetings
     create    app/views/greetings/hello.html.erb
     invoke  test_unit
     create    test/controllers/greetings_controller_test.rb
     invoke  helper
     create    app/helpers/greetings_helper.rb
     invoke  assets
     invoke    coffee
     create      app/assets/javascripts/greetings.coffee
     invoke    scss
     create      app/assets/stylesheets/greetings.scss
```
We can see all the files that are generated for us. To link the `greetings.scss` inside our `hello.html.erb` file, we need to include:
```
<%= stylesheet_link_tag "greetings" %>
```
if we need to include multiple css files:
```
<%= stylesheet_link_tag "first_css", "second_css" %>
```
or if we need to include css file link:
```
<%= stylesheet_link_tag "http://example.com/main.css" %>
```
often we would see:
```
<%= stylesheet_link_tag "greetings", media: "all" %>
```
`media: "all"` is a css attributes for media types. Rails defaults media types to "screen" which means the styling is intented primarily for color computer screens, thus passing in `all` means the CSS styling is suitable for all devices.

If we are not using `rails generate` command, we can also include those files needed manually.

2. `application.css` is meant for application wide styles as it is linked to `application.html.erb` inside `app/views/layouts`. For other CSS styling it can be added using method mentioned above or manually. It is generally better to create a new file per style scope.

tlkiong replied:
1. Ok.
2. Wrong. application.css is the entry point where sprockets find for all connected CSS files to be loaded. You can have application wide CSS in other files as well.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/channels/application_cable/channel.rb:
```diff
@@ -0,0 +1,4 @@
+module ApplicationCable
```
what is this file for?

Owlullaby replied:
`channel.rb` inherits from ActionCable::Channel::Base and it's responsible for pushing contents to the subscribers of the channel.

tlkiong replied:
This kind of answer shows me that you jstu copy paste what you read without understanding it. If you going to learn like this, then better not waste your time

How is it responsible to push content? Where does it does the pushing? how does it do the pushing? magically?

Owlullaby replied:
When `consumer` subscribes to a channel, they are called the `subscriber`. Action Cable uses Pub/Sub or Publish-Subscribe approach to communicate between server and clients. In a Pub/Sub system, the publisher send contents to the channel and those subscribers that subscribed to that specific channel would receive the the contents from the publisher. Some examples of a pub/sub usage is to deliver contents to users' feed based on the topic they subscribed in a content delivering system or in a chat system where users in a same room would be in a same channel.

tlkiong replied:
you are now describing what Pub/Sub & Action Cable is. You still haven't explained what is this file for and how it works.

Think of it like: Everything you mention is theory. How you going to implement it?

Owlullaby replied:
Rails by default creats a parent `ApplicationCable::Channel` class for encapsulating shared logic between your channels which is the `channels/application_cable/channel.rb` file.
You can then create additional files in `/channels` for example `chat.rb` as below for different channels for consumers to subscribe to.
```
# /application_cable/chat.rb

class Chat < ApplicationCable::Channel
end
```

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/channels/application_cable/connection.rb:
```diff
@@ -0,0 +1,4 @@
+module ApplicationCable
```
what is this file for?

Owlullaby replied:
`connection.rb` inherits from ActionCable::Connection::Base and it's responsible for authentication and authorization of the connections from current users.

tlkiong replied:
This kind of answer shows me that you jstu copy paste what you read without understanding it. If you going to learn like this, then better not waste your time

How is it responsible for authentication and authorization? Where does it does that? How does it does that? Magically?

Whats the difference between authorization & authentication?

Owlullaby replied:
Authentication using [Devise](https://github.com/heartcombo/devise)
Adding the code below in `connection.rb`:
```
module ApplicationCable
  class Connection < ActionCable::Connection::Base
    identified_by :current_user

    def connect
      self.current_user = find_verified_user
      logger.add_tags 'ActionCable', current_user.email
    end

    protected

    def find_verified_user # this checks whether a user is authenticated with devise
      if verified_user = env['warden'].user
        verified_user
      else
        reject_unauthorized_connection
      end
    end
  end
end

```
`authentication` is the process of verifying the identity of the consumers while `authorization` authorizes the consumers access to specific part of the sites.

tlkiong replied:
1. You are showing me EXTRA code that you ADD in yourself to do authentication by using the devise gem. This file is NOT responsible for doing it. Let me prove it to you. Move the entire code for authentication out of this file and put it in another file. Then, this file now does not have any authentication. So, I ask you again. What does this file have to do with being responsible for authentication & authorization?

Owlullaby replied:
Rails by default creates a `ApplicationCable::Connection` class inside the `application_cable/connection.rb` file. For each WebSocket connection established, a connection object is created which becomes the parent for all channel subscriptions that are created after on.
>How is it responsible for authentication and authorization? Where does it does that? How does it does that? Magically?

I thought you mean by how to implement it inside the file, hence the code, then recent comment only I realized that you mean that it wasn't meant for authentication and authorization as default.

tlkiong replied:
> For each WebSocket connection established, a connection object is created which becomes the parent for all channel subscriptions that are created after on.

What object & how?

Owlullaby replied:
Connection object is an instance of `ApplicationCable::Connection` which inherits `ActionCable::Connection::Base`.

On the client-side, consumers require an instance of the connection and a consumer becomes a subscriber by creating a subscription to a channel. 

Rails by default creates the following file `app/javascript/channels/consumer.js`
```
// app/javascript/channels/consumer.js

import { createConsumer } from "@rails/actioncable"

export default createConsumer()

```
To subscribe to a specific channel:
```
// app/javascript/channels/chat_channel.js

import consumer from "./consumer"

consumer.subscriptions.create({ channel: "ChatChannel" })
```

tlkiong replied:
Still didn't answer my question.

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/controllers/application_controller.rb:
```diff
@@ -0,0 +1,2 @@
+class ApplicationController < ActionController::Base
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/helpers/application_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module ApplicationHelper
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/javascript/channels/consumer.js:
```diff
@@ -0,0 +1,6 @@
+// Action Cable provides the framework to deal with WebSockets in Rails.
```
what is this file for?

Owlullaby replied:
`consumer` is the client of a WebSocket connection and consumer is created by the client-side JavaScript framework. In rails, the `consumer.js` file is created on default to create an instance of the WebSocket connection.

tlkiong replied:
what is `client`?
What if you don't have your frontend on rails but on others. How do you magically connect to this?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/javascript/channels/index.js:
```diff
@@ -0,0 +1,5 @@
+// Load all the channels within this directory and all subdirectories.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/javascript/packs/application.js:
```diff
@@ -0,0 +1,13 @@
+// This file is automatically compiled by Webpack, along with any other files
```
what is this file for?

Owlullaby replied:
`application.js` is where you include all the Javascript scripts that you needed for your site.

tlkiong replied:
1. `include` How?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/jobs/application_job.rb:
```diff
@@ -0,0 +1,7 @@
+class ApplicationJob < ActiveJob::Base
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/mailers/application_mailer.rb:
```diff
@@ -0,0 +1,4 @@
+class ApplicationMailer < ActionMailer::Base
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/models/application_record.rb:
```diff
@@ -0,0 +1,3 @@
+class ApplicationRecord < ActiveRecord::Base
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/views/layouts/application.html.erb:
```diff
@@ -0,0 +1,16 @@
+<!DOCTYPE html>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/views/layouts/mailer.html.erb:
```diff
@@ -0,0 +1,13 @@
+<!DOCTYPE html>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/app/views/layouts/mailer.text.erb:
```diff
@@ -0,0 +1 @@
+<%= yield %>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/bundle:
```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/rails:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/rake:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/setup:
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/spring:
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/bin/yarn:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env ruby
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config.ru:
```diff
@@ -0,0 +1,6 @@
+# This file is used by Rack-based servers to start the application.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/application.rb:
```diff
@@ -0,0 +1,22 @@
+require_relative "boot"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/boot.rb:
```diff
@@ -0,0 +1,4 @@
+ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../Gemfile', __dir__)
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/cable.yml:
```diff
@@ -0,0 +1,10 @@
+development:
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/credentials.yml.enc:
```diff
@@ -0,0 +1 @@
+lk8+G7IeT4D58hmBwYWdWIY1PGA9fmgCkGtLaCuzap+FC9zHqcQVHcmaH+hvPh23RBPKZ9HnJv/iRBROq8WwmikriKGvF5tplhcPVtwxef63+tpsdPZguj53QljUXH7NDSo3nyl6lmvwmONS5OOgXZPxzm/Zid2zRy3FCOZc3stV33+Qk3u+UHQaTNpSrEAq1neQZXJsPdoayn0rnt/YebKANt8yU4/W+wkTLfHspdNbwxZXaaXbxKym3gy4QZ3XJyf6Nqz6Bm0kp1qOixvkToK476AYK24dFoebMy1kG9KBqlDDRcmkBbkoh0LZUqxAb3SqJGto9IJBqShPWGvkwBFspJzAwJZolnrrxLGpt1ofd56v/C1paX+tfhKgjRRlI8DednfUlKHsKiF ...
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/database.yml:
```diff
@@ -0,0 +1,25 @@
+# SQLite. Versions 3.8.0 and up are supported.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/environment.rb:
```diff
@@ -0,0 +1,5 @@
+# Load the Rails application.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/environments/development.rb:
```diff
@@ -0,0 +1,76 @@
+require "active_support/core_ext/integer/time"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/environments/production.rb:
```diff
@@ -0,0 +1,120 @@
+require "active_support/core_ext/integer/time"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/environments/test.rb:
```diff
@@ -0,0 +1,60 @@
+require "active_support/core_ext/integer/time"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/application_controller_renderer.rb:
```diff
@@ -0,0 +1,8 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/assets.rb:
```diff
@@ -0,0 +1,14 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/backtrace_silencers.rb:
```diff
@@ -0,0 +1,8 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/content_security_policy.rb:
```diff
@@ -0,0 +1,30 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/cookies_serializer.rb:
```diff
@@ -0,0 +1,5 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/filter_parameter_logging.rb:
```diff
@@ -0,0 +1,6 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/mime_types.rb:
```diff
@@ -0,0 +1,4 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/initializers/wrap_parameters.rb:
```diff
@@ -0,0 +1,14 @@
+# Be sure to restart your server when you modify this file.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/locales/en.yml:
```diff
@@ -0,0 +1,33 @@
+# Files in the config/locales directory are used for internationalization
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/puma.rb:
```diff
@@ -0,0 +1,43 @@
+# Puma can serve each request in a thread from an internal thread pool.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/routes.rb:
```diff
@@ -0,0 +1,3 @@
+Rails.application.routes.draw do
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/spring.rb:
```diff
@@ -0,0 +1,6 @@
+Spring.watch(
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/config/storage.yml:
```diff
@@ -0,0 +1,34 @@
+test:
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/db/seeds.rb:
```diff
@@ -0,0 +1,7 @@
+# This file should contain all the record creation needed to seed the database with its default values.
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/package.json:
```diff
@@ -0,0 +1,11 @@
+{
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/public/404.html:
```diff
@@ -0,0 +1,67 @@
+<!DOCTYPE html>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/public/422.html:
```diff
@@ -0,0 +1,67 @@
+<!DOCTYPE html>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/public/500.html:
```diff
@@ -0,0 +1,66 @@
+<!DOCTYPE html>
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/public/robots.txt:
```diff
@@ -0,0 +1 @@
+# See https://www.robotstxt.org/robotstxt.html for documentation on how to use the robots.txt file
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/test/application_system_test_case.rb:
```diff
@@ -0,0 +1,5 @@
+require "test_helper"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/test/channels/application_cable/connection_test.rb:
```diff
@@ -0,0 +1,11 @@
+require "test_helper"
```
what is this file for?

tlkiong commented on Section 1: Introduction and Setup/4-RoR Development Environment/testing_project/test/test_helper.rb:
```diff
@@ -0,0 +1,13 @@
+ENV['RAILS_ENV'] ||= 'test'
```
what is this file for?

tlkiong reviewed (changes requested):
1. What trello card is this tied to?
2. Make sure your file path has no space in the entire path & all lowercase. Eg: `introduction-and-setup`. Not `Introduction and setup`. Please tell me why I ask you to do this.
3. Please use asdf & `.tool-version`. Remove `.ruby-version` file. Please tell me why I ask to do this
4. Explain why I ask to use conventional commits

---

Note:
When you answer the question, don't google and just copy paste the answer. Understand whats the file about, why they are there (in that exact location) and how they are used

Owlullaby committed:
fix: change file path naming and added .tool-versions

Owlullaby committed:
fix: remove changed directories

Owlullaby commented:
> 1. What trello card is this tied to?
> 
>     2. Make sure your file path has no space in the entire path & all lowercase. Eg: `introduction-and-setup`. Not `Introduction and setup`. Please tell me why I ask you to do this.
> 
>     3. Please use asdf & `.tool-version`. Remove `.ruby-version` file. Please tell me why I ask to do this
> 
>     4. Explain why I ask to use conventional commits
> 
> 
> Note: When you answer the question, don't google and just copy paste the answer. Understand whats the file about, why they are there (in that exact location) and how they are used

1. Added trello card link inside description.
2. Changed directories and files naming path, removing spaces in paths make it easier to navigate through files in the terminal and also when using `git add` to add files to stage.
3. Changed from using `rbenv` to using `asdf`. `asdf` allows managing different tools/ languages versions using one version manager inside a single project, useful for web development when we have multiple versions of tools to keep track of.
4. Using conventional commits to have a defined structure for others to understand well and making contributions to the project easily.

tlkiong commented:
1. Change it to a hyperlink please.
2. `make it easier to navigate` how does making the said changes make it easier to navigate or add files to stage?
3. Ok. So, why use asdf?
4. Not only that, that is because it is a standard. Being a standard means its a consistent thing. Which means people can make tools following that standard.

Owlullaby commented:
> 1. Change it to a hyperlink please.
> 2. `make it easier to navigate` how does making the said changes make it easier to navigate or add files to stage?
> 3. Ok. So, why use asdf?
> 4. Not only that, that is because it is a standard. Being a standard means its a consistent thing. Which means people can make tools following that standard.

1. Done.
2. `make it easier to navigate` in the sense that for me myself while changing directories or managing the files using the terminal using `cd`, `mv`, `mkdir` and such for example if just type in terminal 
```
mkdir introduction and setup
```
It's gonna be creating three folders named `introduction`, `and` and `setup`. I would have to add in a single quote or a double quote to it to make it correct.
```
mkdir 'introduction and setup'
```
And while changing directories, I would need to rely on `tab` to auto finish the file name for me. In a way, it's much more convenient for me if using `-` to replace the spacing.
3. As mentioned, using `asdf` allows managing different tools/ languages versions using one version manager inside a single project. For example, we have different version managers for each languages or tools like `rbenv` for Ruby, `nvm` for Node, `pyenv` for Python and so on. When we are using multiple languages or tools inside our project it gets hefty and time consuming to manage them all through different version manager as compared to using `asdf`, a single version manager to manage them all. 
Inside the root directory of our project, just add a `.tool-versions` file and list down the name of each dependencies with it's version, like the example below:
```
# .tool-versions
ruby 2.7.0
node 12.8

```
Also, when working on the same project on a different machine or when you have others that's trying to run your code, just run `asdf install` from the directory to have all the necessary version of each dependencies ready to be used.
4. Okay, thanks for the add.

Owlullaby committed:
fix: adding unwantedfile to gitignore
